CI/CD: Windows test to include MS SQL

Had to remove mssql from CI because it kept failing:

```
FAILED qatrack/reports/tests/test_base.py::TestSaveReport::test_form_valid_with_valid_report_form - django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!
1 failed, 743 passed, 1 skipped, 32 deselected, 238 warnings in 239.48s (0:03:59)
```

Issue appears to need some evaluation, leaving for a seperate PR.

Run uv run pytest -x -m "not selenium"
..............................s......................................... [  6%]
........................................................................ [ 13%]
........................................................................ [ 20%]
........................................................................ [ 27%]
........................................................................ [ 34%]
........................................................................ [ 40%]
........................................................................ [ 47%]
........................................................................ [ 54%]
........................................................................ [ 61%]
........................................................................ [ 68%]
........................F
================================== FAILURES ===================================
____________ TestSaveReport.test_form_valid_with_valid_report_form ____________

self = <django.db.backends.utils.CursorWrapper object at 0x0000022BDF9934A0>
sql = 'SET NOCOUNT ON INSERT INTO [reports_savedreport] ([title], [report_type], [report_format], [filters], [include_signat...ified], [modified_by_id]) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s); SELECT CAST(SCOPE_IDENTITY() AS bigint)'
params = ('Title', 'testlistinstance_summary', 'pdf', '"{\\"work_completed\\": [\\"2000-01-01T00:00:00-05:00\\", \\"2000-02-01T...it\\": [], \\"unit_test_collection__frequency\\": [], \\"unit_test_collection__assigned_to\\": []}"', True, False, ...)
ignored_wrapper_args = (False, {'connection': <DatabaseWrapper vendor='microsoft' alias='default'>, 'cursor': <django.db.backends.utils.CursorWrapper object at 0x0000022BDF9934A0>})

    def _execute(self, sql, params, *ignored_wrapper_args):
        self.db.validate_no_broken_transaction()
        with self.db.wrap_database_errors:
            if params is None:
                # params default might be backend specific.
                return self.cursor.execute(sql)
            else:
>               return self.cursor.execute(sql, params)

.venv\Lib\site-packages\django\db\backends\utils.py:89: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <mssql.base.CursorWrapper object at 0x0000022BDF993080>
sql = 'SET NOCOUNT ON INSERT INTO [reports_savedreport] ([title], [report_type], [report_format], [filters], [include_signat...y_id], [modified], [modified_by_id]) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?); SELECT CAST(SCOPE_IDENTITY() AS bigint)'
params = ('Title', 'testlistinstance_summary', 'pdf', '"{\\"work_completed\\": [\\"2000-01-01T00:00:00-05:00\\", \\"2000-02-01T...ion__unit\\": [], \\"unit_test_collection__frequency\\": [], \\"unit_test_collection__assigned_to\\": []}"', 1, 0, ...)

    def execute(self, sql, params=None):
        self.last_sql = sql
        if 'GROUP BY' in sql:
            sql, params = self.format_group_by_params(sql, params)
        sql = self.format_sql(sql, params)
        params = self.format_params(params)
        self.last_params = params
        try:
>           return self.cursor.execute(sql, params)
E           pyodbc.IntegrityError: ('23000', '[23000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]The INSERT statement conflicted with the CHECK constraint "reports_savedreport_filters_a01ea7f1_check". The conflict occurred in database "test_qatrackplus31", table "dbo.reports_savedreport", column \'filters\'. (547) (SQLExecDirectW); [23000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]The statement has been terminated. (3621)')

.venv\Lib\site-packages\mssql\base.py:773: IntegrityError

The above exception was the direct cause of the following exception:

request = <WSGIRequest: POST '/reports/save/'>

    @wraps(get_response)
    def inner(request):
        try:
>           response = get_response(request)

.venv\Lib\site-packages\django\core\handlers\exception.py:55: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv\Lib\site-packages\django\core\handlers\base.py:197: in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
.venv\Lib\site-packages\django\views\decorators\http.py:43: in inner
    return func(request, *args, **kwargs)
qatrack\reports\views.py:157: in save_report
    saved_report.save()
.venv\Lib\site-packages\django\db\models\base.py:814: in save
    self.save_base(
.venv\Lib\site-packages\django\db\models\base.py:877: in save_base
    updated = self._save_table(
.venv\Lib\site-packages\django\db\models\base.py:1020: in _save_table
    results = self._do_insert(
.venv\Lib\site-packages\django\db\models\base.py:1061: in _do_insert
    return manager._insert(
.venv\Lib\site-packages\django\db\models\manager.py:87: in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
.venv\Lib\site-packages\django\db\models\query.py:1805: in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
.venv\Lib\site-packages\django\db\models\sql\compiler.py:1820: in execute_sql
    cursor.execute(sql, params)
.venv\Lib\site-packages\django\db\backends\utils.py:67: in execute
    return self._execute_with_wrappers(
.venv\Lib\site-packages\django\db\backends\utils.py:80: in _execute_with_wrappers
    return executor(sql, params, many, context)
.venv\Lib\site-packages\django\db\backends\utils.py:84: in _execute
    with self.db.wrap_database_errors:
.venv\Lib\site-packages\django\db\utils.py:91: in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
.venv\Lib\site-packages\django\db\backends\utils.py:89: in _execute
    return self.cursor.execute(sql, params)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <mssql.base.CursorWrapper object at 0x0000022BDF993080>
sql = 'SET NOCOUNT ON INSERT INTO [reports_savedreport] ([title], [report_type], [report_format], [filters], [include_signat...y_id], [modified], [modified_by_id]) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?); SELECT CAST(SCOPE_IDENTITY() AS bigint)'
params = ('Title', 'testlistinstance_summary', 'pdf', '"{\\"work_completed\\": [\\"2000-01-01T00:00:00-05:00\\", \\"2000-02-01T...ion__unit\\": [], \\"unit_test_collection__frequency\\": [], \\"unit_test_collection__assigned_to\\": []}"', 1, 0, ...)

    def execute(self, sql, params=None):
        self.last_sql = sql
        if 'GROUP BY' in sql:
            sql, params = self.format_group_by_params(sql, params)
        sql = self.format_sql(sql, params)
        params = self.format_params(params)
        self.last_params = params
        try:
>           return self.cursor.execute(sql, params)
E           django.db.utils.IntegrityError: ('23000', '[23000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]The INSERT statement conflicted with the CHECK constraint "reports_savedreport_filters_a01ea7f1_check". The conflict occurred in database "test_qatrackplus31", table "dbo.reports_savedreport", column \'filters\'. (547) (SQLExecDirectW); [23000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]The statement has been terminated. (3621)')

.venv\Lib\site-packages\mssql\base.py:773: IntegrityError

The above exception was the direct cause of the following exception:

self = <test_base.TestSaveReport testMethod=test_form_valid_with_valid_report_form>

    def test_form_valid_with_valid_report_form(self):
        """Invalid base form so should get errors"""
        data = {
            'root-report_type': 'testlistinstance_summary',
            'root-title': 'Title',
            'root-report_format': 'pdf',
            'root-include_signature': True,
            'work_completed': '01 Jan 2000 - 01 Feb 2000',
            'reportnote_set-INITIAL_FORMS': 0,
            'reportnote_set-TOTAL_FORMS': 0,
        }
        assert models.SavedReport.objects.count() == 0
>       payload = self.client.post(self.url, data).json()

qatrack\reports\tests\test_base.py:220: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv\Lib\site-packages\django\test\client.py:948: in post
    response = super().post(
.venv\Lib\site-packages\django\test\client.py:482: in post
    return self.generic(
.venv\Lib\site-packages\django\test\client.py:609: in generic
    return self.request(**r)
.venv\Lib\site-packages\django\test\client.py:886: in request
    response = self.handler(environ)
.venv\Lib\site-packages\django\test\client.py:176: in __call__
    response = self.get_response(request)
.venv\Lib\site-packages\django\core\handlers\base.py:140: in get_response
    response = self._middleware_chain(request)
.venv\Lib\site-packages\django\core\handlers\exception.py:57: in inner
    response = response_for_exception(request, exc)
.venv\Lib\site-packages\django\core\handlers\exception.py:140: in response_for_exception
    response = handle_uncaught_exception(
.venv\Lib\site-packages\django\core\handlers\exception.py:185: in handle_uncaught_exception
    return callback(request)
qatrack\qatrack_core\views.py:220: in handle_500
    return handle_error(
qatrack\qatrack_core\views.py:192: in handle_error
    return render(request, 'site_error.html', context, status=code)
.venv\Lib\site-packages\django\shortcuts.py:24: in render
    content = loader.render_to_string(template_name, context, request, using=using)
.venv\Lib\site-packages\django\template\loader.py:62: in render_to_string
    return template.render(context, request)
.venv\Lib\site-packages\django\template\backends\django.py:61: in render
    return self.template.render(context)
.venv\Lib\site-packages\django\template\base.py:173: in render
    with context.bind_template(self):
C:\hostedtoolcache\windows\Python\3.12.10\x64\Lib\contextlib.py:137: in __enter__
    return next(self.gen)
.venv\Lib\site-packages\django\template\context.py:254: in bind_template
    updates.update(processor(self.request))
qatrack\context_processors.py:152: in site
    cur_site = get_current_site(request)
.venv\Lib\site-packages\django\contrib\sites\shortcuts.py:16: in get_current_site
    return Site.objects.get_current(request)
.venv\Lib\site-packages\django\contrib\sites\models.py:59: in get_current
    return self._get_site_by_id(site_id)
.venv\Lib\site-packages\django\contrib\sites\models.py:30: in _get_site_by_id
    site = self.get(pk=site_id)
.venv\Lib\site-packages\django\db\models\manager.py:87: in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
.venv\Lib\site-packages\django\db\models\query.py:633: in get
    num = len(clone)
.venv\Lib\site-packages\django\db\models\query.py:380: in __len__
    self._fetch_all()
.venv\Lib\site-packages\django\db\models\query.py:1881: in _fetch_all
    self._result_cache = list(self._iterable_class(self))
.venv\Lib\site-packages\django\db\models\query.py:91: in __iter__
    results = compiler.execute_sql(
.venv\Lib\site-packages\django\db\models\sql\compiler.py:1560: in execute_sql
    cursor.execute(sql, params)
.venv\Lib\site-packages\django\db\backends\utils.py:67: in execute
    return self._execute_with_wrappers(
.venv\Lib\site-packages\django\db\backends\utils.py:80: in _execute_with_wrappers
    return executor(sql, params, many, context)
.venv\Lib\site-packages\django\db\backends\utils.py:83: in _execute
    self.db.validate_no_broken_transaction()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <DatabaseWrapper vendor='microsoft' alias='default'>

    def validate_no_broken_transaction(self):
        if self.needs_rollback:
>           raise TransactionManagementError(
                "An error occurred in the current transaction. You can't "
                "execute queries until the end of the 'atomic' block."
            ) from self.rollback_exc
E           django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.

.venv\Lib\site-packages\django\db\backends\base\base.py:531: TransactionManagementError
============================== warnings summary ===============================
.venv\Lib\site-packages\django\conf\__init__.py:267
  D:\a\qatrackplus\qatrackplus\.venv\Lib\site-packages\django\conf\__init__.py:267: RemovedInDjango50Warning: The USE_L10N setting is deprecated. Starting with Django 5.0, localized formatting of data will always be enabled. For example Django will display numbers and dates using the format of the current locale.
    warnings.warn(USE_L10N_DEPRECATED_MSG, RemovedInDjango50Warning)

.venv\Lib\site-packages\pynliner\soupselect.py:159
  D:\a\qatrackplus\qatrackplus\.venv\Lib\site-packages\pynliner\soupselect.py:159: SyntaxWarning: invalid escape sequence '\.'
    classes = re.findall('\.([a-zA-Z0-9_-]+)', token)

qatrack\attachments\models.py:1
  D:\a\qatrackplus\qatrackplus\qatrack\attachments\models.py:1: DeprecationWarning: 'imghdr' is deprecated and slated for removal in Python 3.13
    import imghdr

qatrack\qatrack_core\tests\live.py:92
  D:\a\qatrackplus\qatrackplus\qatrack\qatrack_core\tests\live.py:92: PytestUnknownMarkWarning: Unknown pytest.mark.selenium - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.selenium

qatrack\qa\tests\test_selenium.py:170
  D:\a\qatrackplus\qatrackplus\qatrack\qa\tests\test_selenium.py:170: PytestUnknownMarkWarning: Unknown pytest.mark.selenium - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.selenium

qatrack\qa\tests\test_selenium.py:576
  D:\a\qatrackplus\qatrackplus\qatrack\qa\tests\test_selenium.py:576: PytestUnknownMarkWarning: Unknown pytest.mark.selenium - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.selenium

qatrack\qa\tests\test_selenium.py:896
  D:\a\qatrackplus\qatrackplus\qatrack\qa\tests\test_selenium.py:896: PytestUnknownMarkWarning: Unknown pytest.mark.selenium - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.selenium

qatrack\qa\tests\test_tags.py:154
  D:\a\qatrackplus\qatrackplus\qatrack\qa\tests\test_tags.py:154: RemovedInDjango50Warning: The django.utils.timezone.utc alias is deprecated. Please update your code to use datetime.timezone.utc instead.
    dtstart=timezone.datetime(2012, 1, 1, tzinfo=timezone.utc),

qatrack\qatrack_core\tests\test_forms.py:7
  D:\a\qatrackplus\qatrackplus\qatrack\qatrack_core\tests\test_forms.py:7: PytestCollectionWarning: cannot collect test class 'TestForm' because it has a __init__ constructor (from: qatrack/qatrack_core/tests/test_forms.py)
    class TestForm(BetterFormMixin, forms.Form):

qatrack/accounts/tests/test_accounts.py::TestAdminFilter::test_lookups
  D:\a\qatrackplus\qatrackplus\qatrack\qa\utils.py:196: RemovedInDjango51Warning: BaseUserManager.make_random_password() is deprecated.
    pwd = make_password(user_klass.objects.make_random_password())

qatrack/accounts/tests/test_accounts.py::TestAdminFilter::test_lookups
  D:\a\qatrackplus\qatrackplus\qatrack\accounts\migrations\0001_initial.py:10: RemovedInDjango51Warning: BaseUserManager.make_random_password() is deprecated.
    user.password = make_password(User.objects.make_random_password())

qatrack/api/qa/tests/test_api.py: 69 warnings
qatrack/qa/tests/test_admin.py: 34 warnings
qatrack/qa/tests/test_models.py: 1 warning
qatrack/qa/tests/test_views.py: 43 warnings
  D:\a\qatrackplus\qatrackplus\.venv\Lib\site-packages\django\db\models\fields\related_lookups.py:52: RemovedInDjango50Warning: Passing unsaved model instances to related filters is deprecated.
    warnings.warn(

qatrack/qa/tests/test_admin_views.py: 1 warning
qatrack/qa/tests/test_utils.py: 12 warnings
  D:\a\qatrackplus\qatrackplus\qatrack\qa\testpack.py:69: RemovedInDjango50Warning: The django.utils.timezone.utc alias is deprecated. Please update your code to use datetime.timezone.utc instead.
    'datetime': "%s" % (timezone.now().astimezone(timezone.utc)),

qatrack/qa/tests/test_models.py::TestUnitTestCollection::test_date_straddle_due_status
qatrack/qa/tests/test_models.py::TestUnitTestCollection::test_date_straddle_due_status
qatrack/qa/tests/test_models.py::TestUnitTestCollection::test_date_straddle_due_status
qatrack/qa/tests/test_models.py::TestUnitTestCollection::test_date_straddle_due_status
qatrack/qa/tests/test_models.py::TestUnitTestCollection::test_date_straddle_due_status
  D:\a\qatrackplus\qatrackplus\qatrack\qa\tests\test_models.py:21: RemovedInDjango50Warning: The django.utils.timezone.utc alias is deprecated. Please update your code to use datetime.timezone.utc instead.
    return timezone.make_aware(timezone.datetime(2014, 4, 2, 2), timezone.utc)

qatrack/qa/tests/test_views.py: 52 warnings
  D:\a\qatrackplus\qatrackplus\qatrack\qa\views\charts.py:276: RemovedInDjango50Warning: The django.utils.timezone.utc alias is deprecated. Please update your code to use datetime.timezone.utc instead.
    return [d_from.astimezone(timezone.utc), d_to.astimezone(timezone.utc)]

qatrack/qa/tests/test_views.py::TestURLS::test_qa_urls
qatrack/qa/tests/test_views.py::TestReviewTestListInstance::test_review_tli_url
qatrack/qa/tests/test_views.py::TestReviewTestListInstance::test_review_tli_url
qatrack/qa/tests/test_views.py::TestReviewTestListInstance::test_review_tli_url
qatrack/qa/tests/test_views.py::TestReviewTestListInstance::test_review_tli_url
qatrack/qa/tests/test_views.py::TestReviewTestListInstance::test_update
qatrack/qa/tests/test_views.py::TestReviewTestListInstance::test_update_still_requires_review
  D:\a\qatrackplus\qatrackplus\.venv\Lib\site-packages\django\template\base.py:1047: RemovedInDjango50Warning: The "default.html" templates for forms and formsets will be removed. These were proxies to the equivalent "table.html" templates, but the new "div.html" templates will be the default from Django 5.0. Transitional renderers are provided to allow you to opt-in to the new output style now. See https://docs.djangoproject.com/en/4.2/releases/4.1/ for more details
    value = str(value)

qatrack/qatrack_core/tests/test_zoneinfo.py::TestZoneinfoDateUtilities::test_local_end_of_day
  D:\a\qatrackplus\qatrackplus\qatrack\qatrack_core\tests\test_zoneinfo.py:108: RemovedInDjango50Warning: The django.utils.timezone.utc alias is deprecated. Please update your code to use datetime.timezone.utc instead.
    dt = timezone.datetime(2023, 6, 15, 14, 30, 45, tzinfo=timezone.utc)

qatrack/qatrack_core/tests/test_zoneinfo.py::TestZoneinfoDateUtilities::test_local_start_of_day
  D:\a\qatrackplus\qatrackplus\qatrack\qatrack_core\tests\test_zoneinfo.py:96: RemovedInDjango50Warning: The django.utils.timezone.utc alias is deprecated. Please update your code to use datetime.timezone.utc instead.
    dt = timezone.datetime(2023, 6, 15, 14, 30, 45, tzinfo=timezone.utc)

qatrack/qatrack_core/tests/test_zoneinfo.py::TestCrossTimezoneCompatibility::test_utc_conversion_consistency
  D:\a\qatrackplus\qatrackplus\qatrack\qatrack_core\tests\test_zoneinfo.py:308: RemovedInDjango50Warning: The django.utils.timezone.utc alias is deprecated. Please update your code to use datetime.timezone.utc instead.
    utc_dt = local_dt.astimezone(timezone.utc)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ===========================
SKIPPED [1] qatrack\api\qa\tests\test_api.py:366: causing a segfault with sqlite :(
FAILED qatrack/reports/tests/test_base.py::TestSaveReport::test_form_valid_with_valid_report_form - django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!
1 failed, 743 passed, 1 skipped, 32 deselected, 238 warnings in 268.87s (0:04:28)
Error: Process completed with exit code 1.

```
FAILED qatrack/reports/tests/test_base.py::TestSaveReport::test_form_valid_with_valid_report_form - django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
```

is this a timing issue?

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -147,7 +147,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        db: [sqlite]
+        db: [sqlite, mssql]
     
     steps:
       - uses: actions/checkout@34e114876b0b11c390a56381ad16ebd13914f8d5

--- a/qatrack/qatrack_core/tasks.py
+++ b/qatrack/qatrack_core/tasks.py
@@ -92,35 +92,54 @@ def run_periodic_scheduler(model, log_name, handler, time_field="time", recurren
     logger.info("Running %s task at %s for notices between %s and %s" % (log_name, now, start_time, end_time))
 
     # first narrow down notices to those supposed to run in this 15 minute block
-    instances = model.objects.filter(
-        **{
-            "%s__gte" % time_field: start_time.strftime("%H:%M"),
-            "%s__lte" % time_field: end_time.strftime("%H:%M"),
-        }
-    )
+    start_time_str = start_time.strftime("%H:%M")
+    end_time_str = end_time.strftime("%H:%M")
+
+    if start_time_str <= end_time_str:
+        instances = model.objects.filter(
+            **{
+                "%s__gte" % time_field: start_time_str,
+                "%s__lte" % time_field: end_time_str,
+            }
+        )
+    else:
+        from django.db.models import Q
+        instances = model.objects.filter(
+            Q(**{"%s__gte" % time_field: start_time_str}) |
+            Q(**{"%s__lte" % time_field: end_time_str})
+        )
 
     if settings.DEBUG:  # pragma: nocover
         instances = model.objects.all()
 
+    tz = ZoneInfo(settings.TIME_ZONE)
+
     # now look at recurrences occuring today and see if they should be sent now
     for instance in instances:
+        t = getattr(instance, time_field)
 
-        # since our recurrence can only occur with a maximum frequency of 1 /
-        # day, between will just return ~timezone.now() if the report is to be
-        # sent today.  If we make reports available at a higher frequency this
-        # check will need to be adjusted.
-        # note: inc=True required to catch daily recurrences
-        occurences = getattr(instance, recurrence_field).between(start_today, end_today, inc=True)
-        logger.info(
-            "Occurences for %s %s: %s (between %s & %s)" % (
-                model._meta.model_name,
-                instance.id,
-                occurences,
-                start_today,
-                end_today,
-            )
-        )
-        if occurences and start_time.time() <= getattr(instance, time_field) <= end_time.time():
-            tz = ZoneInfo(settings.TIME_ZONE)
-            send_time = timezone.datetime.combine(start_today, getattr(instance, time_field)).replace(tzinfo=tz)
-            handler(instance, send_time)
+        for day_offset in (0, 1):
+            target_date = start_today.date() + timezone.timedelta(days=day_offset)
+            dt = timezone.datetime.combine(target_date, t)
+
+            if start_time <= dt <= end_time:
+                target_start = start_today + timezone.timedelta(days=day_offset)
+                target_end = end_today + timezone.timedelta(days=day_offset)
+
+                occurences = getattr(instance, recurrence_field).between(target_start, target_end, inc=True)
+                
+                logger.info(
+                    "Occurences for %s %s: %s (between %s & %s)" % (
+                        model._meta.model_name,
+                        instance.id,
+                        occurences,
+                        target_start,
+                        target_end,
+                    )
+                )
+
+                if occurences:
+                    send_time = dt.replace(tzinfo=tz)
+                    handler(instance, send_time)
+                
+                break

--- a/qatrack/qatrack_core/tasks.py
+++ b/qatrack/qatrack_core/tasks.py
@@ -5,6 +5,7 @@ from zoneinfo import ZoneInfo
 
 from django.conf import settings
 from django.db import ProgrammingError, connection
+from django.db.models import Q
 from django.utils import timezone
 from django_q.models import Schedule
 from django_q.tasks import schedule
@@ -103,7 +104,6 @@ def run_periodic_scheduler(model, log_name, handler, time_field="time", recurren
             }
         )
     else:
-        from django.db.models import Q
         instances = model.objects.filter(
             Q(**{"%s__gte" % time_field: start_time_str}) |
             Q(**{"%s__lte" % time_field: end_time_str})
@@ -126,19 +126,19 @@ def run_periodic_scheduler(model, log_name, handler, time_field="time", recurren
                 target_start = start_today + timezone.timedelta(days=day_offset)
                 target_end = end_today + timezone.timedelta(days=day_offset)
 
-                occurences = getattr(instance, recurrence_field).between(target_start, target_end, inc=True)
+                occurrences = getattr(instance, recurrence_field).between(target_start, target_end, inc=True)
                 
                 logger.info(
-                    "Occurences for %s %s: %s (between %s & %s)" % (
+                    "Occurrences for %s %s: %s (between %s & %s)" % (
                         model._meta.model_name,
                         instance.id,
-                        occurences,
+                        occurrences,
                         target_start,
                         target_end,
                     )
                 )
 
-                if occurences:
+                if occurrences:
                     send_time = dt.replace(tzinfo=tz)
                     handler(instance, send_time)
                 

--- a/qatrack/reports/forms.py
+++ b/qatrack/reports/forms.py
@@ -169,7 +169,7 @@ def serialize_form_data(form_data):
 
         data[k] = v
 
-    return json.dumps(data, cls=DjangoJSONEncoder)
+    return json.loads(json.dumps(data, cls=DjangoJSONEncoder))
 
 
 class ReportScheduleForm(forms.ModelForm):

--- a/qatrack/reports/tests/test_base.py
+++ b/qatrack/reports/tests/test_base.py
@@ -688,20 +688,20 @@ class TestSerializeFormData(TestCase):
         g = Group.objects.create(name="group")
         data = {'groups': Group.objects.all()}
         serialized = forms.serialize_form_data(data)
-        assert json.loads(serialized) == {'groups': [g.pk]}
+        assert serialized == {'groups': [g.pk]}
 
     def test_object(self):
         """Model instance should be serialized as instace pk"""
         g = Group.objects.create(name="group")
         data = {'group': g}
         serialized = forms.serialize_form_data(data)
-        assert json.loads(serialized) == {'group': g.pk}
+        assert serialized == {'group': g.pk}
 
     def test_list(self):
         """iterable of objects without pk attribute should be returned as list"""
         data = {'list': ['a', 'b', 'c']}
         serialized = forms.serialize_form_data(data)
-        assert json.loads(serialized) == data
+        assert serialized == data
 
 
 class TestBaseReport(TestCase):

--- a/qatrack/service_log/tests/test_models.py
+++ b/qatrack/service_log/tests/test_models.py
@@ -1,8 +1,10 @@
+import unittest
+
 from django.conf import settings
-from django.db import transaction
+from django.db import connection, transaction
 from django.db.models import ProtectedError
 from django.db.utils import IntegrityError
-from django.test import TestCase, TransactionTestCase
+from django.test import TransactionTestCase
 from django.utils import timezone
 
 from qatrack.accounts.tests.utils import create_group, create_user
@@ -11,13 +13,14 @@ from qatrack.service_log import models as sl_models
 from qatrack.service_log.tests import utils as sl_utils
 
 
-class TestUnitServiceArea(TestCase):
+class TestUnitServiceArea(TransactionTestCase):
 
     def setUp(self):
 
         self.u = qa_utils.create_unit()
         self.sa = sl_utils.create_service_area()
 
+    @unittest.skipIf(connection.vendor == 'microsoft', "mssql-django does not raise IntegrityError for FK-only unique_together constraints")
     def test_unique_together(self):
 
         sl_utils.create_unit_service_area(unit=self.u, service_area=self.sa)
@@ -31,7 +34,7 @@ class TestUnitServiceArea(TestCase):
         self.assertTrue(self.u.name in str(usa) and self.sa.name in str(usa))
 
 
-class TestServiceEventStatus(TestCase):
+class TestServiceEventStatus(TransactionTestCase):
 
     def setUp(self):
         self.ses = sl_utils.create_service_event_status()
@@ -73,8 +76,9 @@ class TestServiceEventStatus(TestCase):
         self.assertTrue(self.ses.name in str(self.ses))
 
 
-class TestThirdParty(TestCase):
+class TestThirdParty(TransactionTestCase):
 
+    @unittest.skipIf(connection.vendor == 'microsoft', "mssql-django does not raise IntegrityError for FK-only unique_together constraints")
     def test_unique_together(self):
 
         v_01 = qa_utils.create_vendor()
@@ -97,6 +101,7 @@ class TestServiceEventAndRelated(TransactionTestCase):
     def setUp(self):
         self.se = sl_utils.create_service_event()
 
+    @unittest.skipIf(connection.vendor == 'microsoft', "mssql-django does not raise IntegrityError for FK-only unique_together constraints")
     def test_third_party_and_hours(self):
 
         se = sl_models.ServiceEvent.objects.first()
@@ -117,6 +122,7 @@ class TestServiceEventAndRelated(TransactionTestCase):
         self.assertEqual((tp.__class__, tp.id), (h_01.user_or_thirdparty().__class__, h_01.user_or_thirdparty().id))
         self.assertEqual((u_02.__class__, u_02.id), (h_02.user_or_thirdparty().__class__, h_02.user_or_thirdparty().id))
 
+    @unittest.skipIf(connection.vendor == 'microsoft', "mssql-django does not raise IntegrityError for FK-only unique_together constraints")
     def test_group_linkers(self):
 
         se = sl_models.ServiceEvent.objects.first()

--- a/qatrack/service_log/tests/test_models.py
+++ b/qatrack/service_log/tests/test_models.py
@@ -102,18 +102,25 @@ class TestServiceEventAndRelated(TransactionTestCase):
         self.se = sl_utils.create_service_event()
 
     @unittest.skipIf(connection.vendor == 'microsoft', "mssql-django does not raise IntegrityError for FK-only unique_together constraints")
-    def test_third_party_and_hours(self):
+    def test_hours_unique_together(self):
 
         se = sl_models.ServiceEvent.objects.first()
         tp = sl_utils.create_third_party()
 
-        h_01 = sl_utils.create_hours(service_event=se, third_party=tp, user=None)
+        sl_utils.create_hours(service_event=se, third_party=tp, user=None)
 
         with self.assertRaises(IntegrityError):
             with transaction.atomic():
                 sl_models.Hours.objects.create(
                     service_event=se, third_party=tp, user=None, time=timezone.timedelta(hours=1)
                 )
+
+    def test_third_party_and_hours(self):
+
+        se = sl_models.ServiceEvent.objects.first()
+        tp = sl_utils.create_third_party()
+
+        h_01 = sl_utils.create_hours(service_event=se, third_party=tp, user=None)
 
         u_02 = create_user(is_superuser=False, uname='user_02')
         h_02 = sl_utils.create_hours(service_event=se, user=u_02)
@@ -123,6 +130,24 @@ class TestServiceEventAndRelated(TransactionTestCase):
         self.assertEqual((u_02.__class__, u_02.id), (h_02.user_or_thirdparty().__class__, h_02.user_or_thirdparty().id))
 
     @unittest.skipIf(connection.vendor == 'microsoft', "mssql-django does not raise IntegrityError for FK-only unique_together constraints")
+    def test_group_linkers_unique_together(self):
+
+        se = sl_models.ServiceEvent.objects.first()
+        g_01 = create_group()
+
+        gl_01 = sl_utils.create_group_linker(group=g_01)
+        gl_01_name = gl_01.name
+
+        with self.assertRaises(IntegrityError):
+            with transaction.atomic():
+                sl_models.GroupLinker.objects.create(name=gl_01_name, group=g_01)
+
+        sl_utils.create_group_linker_instance(group_linker=gl_01, service_event=se)
+
+        with self.assertRaises(IntegrityError):
+            with transaction.atomic():
+                sl_models.GroupLinkerInstance.objects.create(group_linker=gl_01, service_event=se)
+
     def test_group_linkers(self):
 
         se = sl_models.ServiceEvent.objects.first()
@@ -132,17 +157,8 @@ class TestServiceEventAndRelated(TransactionTestCase):
         gl_01 = sl_utils.create_group_linker(group=g_01)
         gl_01_name = gl_01.name
 
-        with self.assertRaises(IntegrityError):
-            with transaction.atomic():
-                sl_models.GroupLinker.objects.create(name=gl_01_name, group=g_01)
-
         gl_02 = sl_utils.create_group_linker(group=g_02)
         sl_utils.create_group_linker_instance(group_linker=gl_01, service_event=se)
-
-        with self.assertRaises(IntegrityError):
-            with transaction.atomic():
-                sl_models.GroupLinkerInstance.objects.create(group_linker=gl_01, service_event=se)
-
         sl_utils.create_group_linker_instance(group_linker=gl_02, service_event=se)
 
         self.assertEqual(2, len(se.grouplinkerinstance_set.all()))

